design-decisions/2018-07-25 Attributes as patterns.md

```diff
@@ -0,0 +1,120 @@
+Attributes as Patterns
+======================
+
+Background
+----------
+
+Attributes are used in Kore in order to provide hints for manipulating symbols,
+axioms, sorts and so on.
+When translating from K, the Kore attributes are a
+fairly direct translation of the K ones.
+
+Each Kore backend can have its own set of attributes, and backends must be able
+to ignore unknown attributes.
+
+Examples include the `constructor` attribute which
+specifies that a certain symbol is a constructor, so, say, it
+cannot be unified with different constructors.
+
+Questions asked
+---------------
+
+1. What kind of structure do attributes have?
+1. What syntax should we use?
+1. How can we make sure that attributes have some documentation?
+
+The answer to the first two questions comes from a Slack discussion on
+`runtimeverification/#haskell`(https://runtimeverification.slack.com/messages/C9B9QBF46/) on 2018-17-25
+
+* between @grosu, @bmmoore, @dwightguth, @u--, @njohnwalker, @traiansf, @hreada.
+* Starting at: https://runtimeverification.slack.com/archives/C9B9QBF46/p1532543176000083
+* Ending at: https://runtimeverification.slack.com/archives/C9B9QBF46/p1532562323000104
+
+The answer to the third
+comes from a face-to-face discussion between @traiansf and @virgil-serbanuta.
+
+Attribute structure
+-------------------
+
+Attributes are identified by a name and might have (or not) parameters.
+
+When present, attribute parameters are usually very simple, e.g. the `strict` attribute
+can have a list of argument indexes in which it is strict.
+
+Sometimes we may
+want to refer to user-defined symbols, e.g. we may use the `unit` attribute
+to specify the neutral element for an operation, and we may want to use this
+attribute both for the operation symbol declaration and for the axioms that
+make it a neutral element.
+
+Similarly, we may want to use an `operation`
+attribute with an operation symbol as argument to, say,
+link an associativity axiom with an operation.
+
+However, we can't tell what's the general structure of an attribute.
+
+Attribute syntax
+----------------
+
+We could encode each attribute as a string (name/key) or as a pair of
+strings (key-value).
+However, that does not seem a natural way of encoding
+the structure that they have, and each backend would need its own
+string-parsing library.
+
+*Decision*: We will encode each attribute as a Kore pattern that uses only
+application and string literals, with the top symbol being an application 
```

Comment:
Why not just say "term pattern"?  That's what term patterns are.  @hreada can you please put  a reference to term patterns here?  Also, I don't see why we have to break the general rule and write `1` instead of `\dv("1","Int")`.  It just feels to me that we are breaking a uniform rule to represent things as patterns, that we know works, for no clear benefits.  "More human readable" is probably your argument.  But why should we bother with that in KORE?  We are going to have layers on top of KORE that add syntactic sugar, as part of implementing K in K.  Frontends can choose to generate those if they want to, and then use a general procedure to "desugar" them, instead of just picking some adhoc encoding now and them being stuck with it.

Reply:
Term patterns in matching logic are patterns built with only functional symbols. See Definition 5.21 at page 28 of the paper _Matching Logic_ at this link http://fsl.cs.illinois.edu/FSL/papers/2017/rosu-2017-lmcs/rosu-2017-lmcs-public.pdf.
If in @traiansf's definition, the "top symbol" is always a functional symbol, then what the definition defines a subset of term patterns. (Notice that string literals, i.e. cons lists of characters, are term patterns by definition).

Reply:
According to @hreada, this is a bit more generic. I'm not sure we need that much precision, so I guess I could just use 'term'.